presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java

```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Test
+    public void testEstimatedMemorySizeDouble()
+    {
+        testEstimatedMemorySize(DOUBLE, i -> (double) i, .05);
+    }
+
+    @Test
+    public void testEstimatedMemorySizeLong()
+    {
+        testEstimatedMemorySize(BIGINT, i -> (long) i, .05);
+    }
+
+    @Test
+    public void testEstimatedMemorySizeString()
+    {
+        testEstimatedMemorySize(VARCHAR, i -> "abcdefghijklmnopqrstuvwxyz".substring(0, i.hashCode() % 26), .05);
+    }
+
+    @Test
+    public void testEstimatedMemorySizeBoolean()
+    {
+        testEstimatedMemorySize(BOOLEAN, i -> i % 2 == 0, 0.5);
+    }
+
+    private <T> void testEstimatedMemorySize(Type type, Function<Integer, T> generator, double tolerance)
+    {
+        KllSketchAggregationState state = new KllSketchAggregationState.Single(type);
+        KllSketchAggregationState.SketchParameters parameters = KllSketchAggregationState.getSketchParameters(type);
+        KllItemsSketch<T> sketch = KllItemsSketch.newHeapInstance(parameters.getComparator(), parameters.getSerde());
+        List<Integer> sizes = ImmutableList.of(512, 2048, 4096, 16384);
+        for (int size : sizes) {
+            IntStream.range(0, size).boxed().map(generator).forEach(sketch::update);
+            long trueSize = GraphLayout.parseInstance(sketch).totalSize();
+            state.setSketch(sketch);
+            long estimatedSize = state.getEstimatedSize();
+            // size should be within margin of the actual size
+            double errorPercent = (double) Math.abs(estimatedSize - trueSize) / trueSize;
+            assertTrue(errorPercent < tolerance, String.format("estimated memory size error for sketch stream size %d was > 5%%: %.2f", size, errorPercent));
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> void testSerializer(Type type, Supplier<KllItemsSketch<?>> sketchSupplier, Collection<T> sketchInputs)
+    {
+        KllSketchStateSerializer serializer = new KllSketchStateSerializer(type);
+        KllSketchAggregationState state = new KllSketchAggregationState.Single(type);
+
+        // serialize empty
+        state.setSketch(sketchSupplier.get());
+        BlockBuilder blockBuilder = VARBINARY.createBlockBuilder(null, 1);
+        serializer.serialize(state, blockBuilder);
+        KllSketchAggregationState deserialized = new KllSketchAggregationState.Single(type);
+        serializer.deserialize(blockBuilder.build(), 0, deserialized);
+
+        // serialize with data
+        KllItemsSketch sketch = state.getSketch();
+        KllItemsSketch actual = sketchSupplier.get();
+        sketchInputs.stream().forEach(i -> {
+            sketch.update(i);
+            actual.update(i);
+        });
+        blockBuilder = VARBINARY.createBlockBuilder(null, 1);
+        state.setSketch(sketch);
+        serializer.serialize(state, blockBuilder);
+        Block serializedBlock = blockBuilder.build();
+        serializer.deserialize(serializedBlock, 0, state);
```

Comment:
Can you deserialize into a new empty state object instead of reusing this `state` object ?